napari/_vispy/overlays/scale_bar.py

```diff
@@ -173,7 +173,35 @@ def _on_box_change(self):
 
     def _on_text_change(self):
         """Update text information"""
-        self.node.text.font_size = self.overlay.font_size
+        # update the dpi scale factor to account for screen dpi
+        # because vispy scales pixel height of text by screen dpi
+        if self.node.text.transforms.dpi:
+            # use 96 as the reference dpi
+            dpi_scale_factor = 96 / self.node.text.transforms.dpi
+        else:
+            dpi_scale_factor = 1
+
+        self.node.text.font_size = self.overlay.font_size * dpi_scale_factor
+        # ensure we recalculate the y_offset from the text size when at top of canvas
+        if 'top' in self.overlay.position:
+            self._on_position_change()
+
+    def _on_position_change(self, event=None):
+        # prevent the text from being cut off by shifting down
+        if 'top' in self.overlay.position:
+            # convert font_size to logical pixels as vispy does
+            # 96 dpi is used as the reference dpi
+            font_logical_pix = self.overlay.font_size * 96 / 72
+            # 7 is base value for the default 10 font size
+            self.y_offset = 7 + font_logical_pix
+        else:
+            self.y_offset = 20
+        super()._on_position_change()
```

Comment:
Perhaps this is the intent? 
```suggestion
        self._update_y_offset()
        super()._on_position_change()

    def _update_y_offset(self):
        """Update the y_offset based on the font size and position."""
        if 'top' in self.overlay.position:
            # convert font_size to logical pixels as vispy does
            # 96 dpi is used as the reference dpi
            font_logical_pix = self.overlay.font_size * 96 / 72
            # 7 is base value for the default 10 font size
            self.y_offset = 7 + font_logical_pix
        else:
            self.y_offset = 20
```

Reply:
Well, whatever I have done different just always sets `y_offset` ... which I don't understand, because it should continue to use the same if 'top' functionality... Enjoy my musings after trying to understand why the two `if 'top'`'s lol

Reply:
We'd still need to implement a `_on_position_change` that would also call this `_update_y_offset`, because you want to adjust the relative positioning when the scale bar is repositioned to the top, so something like:
```
def _on_position_change(self, event=None):
        self._update_y_offset()
        super()._on_position_change()

I'm not sure this refactor is worth it? I think if `_on_position_changed` was a busy callback, then it would make it more readable, but as is, it's just adding boiler plate?
Let's see what Lorenzo has in mind as the architect of all this 😄

Reply:
I think it's better to just leave the `_on_position_change` with the `super()` call for now, less messy for the PR, and we can always come back to refactor later! Sorry againf or the confusing architecture without good documentation @TimMonko, I'll have to write up something better soon :)